dash-renderer/src/actions/isAppReady.js

```diff
@@ -22,7 +31,14 @@ export default (layout, paths, targets) => {
         const ready = isReady(component);
 
         if (ready && typeof ready.then === 'function') {
-            promises.push(ready);
+            promises.push(
+                Promise.race([
+                    ready,
+                    rendered.then(
+                        () => document.getElementById(stringifyId(id)) && ready
+                    ),
+                ])
+            );
```

Comment:
For each component, we wait until either its component type is ready OR we've finished the current render loop and it's not in the DOM (if it _is_ in the DOM we chain on `ready` to ensure this condition is moot)

Note that if there are two components of this type, one rendered and the other not, the `Promise.all` below ensures we still wait for the `ready`. That said I should probably add a test for this case...

Reply:
Thinking about this some more, prior to the DOM check approach, I thought about being able to trigger the component’s download from the renderer, extending the existing component plugin, eg:

https://github.com/plotly/dash/blob/dev/@plotly/dash-component-plugins/src/dynamicImport.js#L23

```
    const get = lazy(() => {
        return Promise.resolve(promise()).then(res => {
            setTimeout(async () => {
                await resolve(true);
                state.isReady = true;
            }, 0);

            return res;
        });
    });

    Object.defineProperty(target, ‘_dashprivate_forceLoadComponent’, {
        get: () => get
    });
```

If the renderer comes across a component using the new version of the plugin, it will be able to trigger the component download by itself. I think the usefulness is limited as the component would still need to be rendered in the DOM in order for its calculated props (table’s `derived_*`) or side-effects (graph’s `extendData`) to take effect.

I’m thinking that instead it might be possible to have preprocessing code that lives outside the components themselves doing `props -> props*` transform, with the renderer being responsible for asking for this transform to be run. This would allow these component props to behave consistently whether they are displayed or not. In the grand scheme of things, this is an edge case, but inconsistent behaviors are always annoying to devs :grimacing: 

I wonder if it’s possible to isolate Plotly.js `fiture + extendData -> figure` logic so that it’s not necessary to load only a small fraction of Plotly.js to do that operation.

In all cases, this is totally out of scope for wildcards. Just want to see what other reviewer's thoughts are about this behavior before going ahead and creating an issue for it.

Reply:
In any case, the callbacks with components that are not part of the layout seem to behave better now.

Reply:
I like the idea of pushing component loading up to the renderer, though would need to ponder a little bit if there are performance implications.

Re: props preprocessing - yay for consistency 😁 but I suspect there will still occasionally be props that do depend on real component rendering. An example is the [live figure](https://github.com/plotly/dash-core-components/issues/584) proposal for `dcc.Graph` - you can't know what that's going to be until the graph actually renders and writes things like axis ranges (when these are auto-determined) back into the figure.